.github/workflows/documentation.yml

```diff
@@ -1,110 +1,166 @@
+
+      # Install odoc if the cache is not retrieved
+      - name: opam install odoc
+        run: opam install odoc
+        if: steps.cache-opam.outputs.cache-hit != 'true'
+
+      # Try to upgrade installed packages and fix dependencies if needed,
+      # when the cache is retrieved
       - run: opam upgrade --fixup
-        # if: steps.cache-opam.outputs.cache-hit == 'true'
-      - run: opam install odoc
+        if: steps.cache-opam.outputs.cache-hit == 'true'
+
+      # Make the documentation
       - name: Make OCaml documentation
         run: opam exec -- make odoc
 
+      # Create and upload an artifact `ocaml_doc` that contains the ocaml
+      # documentation in html format.
+      # This is only done if this workflow is triggered in a PR or on the
+      # following branches : next, main
       - name: Upload ocaml documentation
         uses: actions/upload-artifact@v2
         if: github.event_name == 'pull_request' || github.ref == 'refs/heads/next' || github.ref == 'refs/heads/main'
         with:
           name: ocaml_doc
           path: _build/default/_doc/_html/
 
+
+  # On PR, or push on next/main, build the sphinx general documentation
+  # If this build fails, the documentation workflow stops
+  # If it succed, an artifact is made with the generated documentation
+  # This artifact is used in the deploying job
   sphinx_docs:
-    # Build the sphinx documentation
     name: Sphinx documentation
+
     # We only run this if the ocaml documentation build is successful
     needs: ocaml_docs
-    # Sphinx documentation is only builded if a PR is open or when it's pushed on next or main
+
+    # Sphinx documentation is only builded if a PR is open or when it's
+    # pushed on next or main
     if: github.event_name == 'pull_request' || github.ref == 'refs/heads/next' || github.ref == 'refs/heads/main'
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v1
-      - uses: ammaraskar/sphinx-action@master
+      # Checkout the code of the current branch
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      # Build the sphinx documentation
+      # and automatically print any error or warning
+      - name: Build sphinx documentation
+        uses: ammaraskar/sphinx-action@master
         with:
           docs-folder: "docs/sphinx_docs/"
           build-command: "sphinx-build -b html . _build"
 
+      # Create and upload an artifact `sphinx_doc` that contains the sphinx
+      # documentation in html format.
       - name: Upload sphinx documentation
         uses: actions/upload-artifact@v2
         with:
           name: sphinx_doc
           path: docs/sphinx_docs/_build
 
+
+  # For every push on next, retrieve ocaml and sphinx documentation
+  # and publish them on gh-pages branch
   deploy_docs:
-    # Deploy the documentation only on next
     name: Deploy documentation
+
     if: github.ref == 'refs/heads/next'
 
     needs: sphinx_docs
```

Comment:
Even if this works because of transitive dependencies, if possible I think it'd be better to explicitly state that this job needs the `build_docs` job (since one step explicitly downloads an artifact from that job).